state/protocol/badger/snapshot_test.go

```diff
@@ -301,12 +300,68 @@ func TestSealingSegment(t *testing.T) {
 			require.NoError(t, err)
 
 			// sealing segment should be [B2, B3, B4]
-			assert.Len(t, segment, 3)
-			assert.Equal(t, block2.ID(), segment[0].ID())
-			assert.Equal(t, block3.ID(), segment[1].ID())
-			assert.Equal(t, block4.ID(), segment[2].ID())
+			assert.Len(t, segment.Blocks, 3)
+			assert.Equal(t, block2.ID(), segment.Blocks[0].ID())
+			assert.Equal(t, block3.ID(), segment.Blocks[1].ID())
+			assert.Equal(t, block4.ID(), segment.Blocks[2].ID())
+		})
+	})
+
+	// test sealing segment where a block contains a ExecutionReceipt that
+	// references a ExecutionResult contained in a different block.
+	// the sealing segment should contain any ExecutionResults that were
+	// missing from a blocks payload in the SealingSegment.ExecutionResults field
+	// ROOT -> B1(Receipt_A) -> B2[Result_B, Receipt_B, Receipt_A_2] -> B3 -> B4 -> B5 (Seal_B1)
+	// Expected sealing segment: SealingSegment{Blocks[B1, B2, B3, B4, B5] ExecutionResults[Result_A]}
+	t.Run("sealing segment decoupled execution results and receipts", func(t *testing.T) {
+		block1 := unittest.BlockWithParentFixture(head)
+		receipt1 := unittest.ReceiptForBlockFixture(&block1)
+		util.RunWithFollowerProtocolStateAndResults(t, rootSnapshot, []*flow.ExecutionResult{&receipt1.ExecutionResult}, func(db *badger.DB, state *bprotocol.FollowerState) {
+			// simulate scenario where execution result is missing from block payload
+			// SealingSegment() should get result from results db and store it on ExecutionReceipts
+			// field on SealingSegment
+			block1.SetPayload(unittest.PayloadFixture(unittest.WithReceiptsAndNoResults(receipt1)))
+
+			err = state.Extend(context.Background(), &block1)
+
+			block2 := unittest.BlockWithParentFixture(block1.Header)
+			err = state.Extend(context.Background(), &block2)
+			require.NoError(t, err)
+
+			block3 := unittest.BlockWithParentFixture(block2.Header)
+			err = state.Extend(context.Background(), &block3)
+			require.NoError(t, err)
+
+			block4 := unittest.BlockWithParentFixture(block3.Header)
+			receipt2, seal2 := unittest.ReceiptAndSealForBlock(&block2)
+			block4.SetPayload(unittest.PayloadFixture(unittest.WithReceipts(receipt2), unittest.WithSeals(seal2)))
+			err = state.Extend(context.Background(), &block4)
+			require.NoError(t, err)
+
+			// build the block sealing block 1
+			block5 := unittest.BlockWithParentFixture(block4.Header)
+			receipt3, seal1 := unittest.ReceiptAndSealForBlock(&block1)
+			block5.SetPayload(unittest.PayloadFixture(unittest.WithReceipts(receipt3), unittest.WithSeals(seal1)))
+			err = state.Extend(context.Background(), &block5)
+			require.NoError(t, err)
```

Comment:
This is an invalid block structure, because we are sealing block 2 before block 1. A parent block is always finalized and sealed before its child. This wouldn't be enforced in this component, but it's still better to use internally consistent fixtures for tests we expect to pass.

Reply:
:100: